### PR TITLE
Device network connections as components

### PR DESCRIPTION
**Pull request by juliangiebel**
## About the PR 
This PR changes device network Connections to Components.
The following components where added:
* WiredNetworkConnection
* WirelessNetworkConnection

Network connection components use component messaging.

There are also some fixes to the device network and the configuration component.

## Todo:

- [x] Write documentation for both network connection components

### CODE REVIEW

**Review by PaulRitter — CHANGES_REQUESTED** [review]

File: Content.Server/DeviceNetwork/NetworkPacket.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        private NetworkPayload(int size) : base(size)
+        {
+        }
+
+        /// <summary>
+        /// Crates a device network packet, drops invalid key value pairs
+        /// </summary>
+        /// <param name="data">An array of tuples containing key value pairs</param>
+        /// <returns>The created device network packet</returns>
+        public static NetworkPayload Create(params (string, string)[] data)
```
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/DeviceNetwork/NetworkPacket.cs:18]: The goal was to have a dictionary that gets initialized with the correct size. Would it be enough if i ignore duplicate keys and log an error?

[the remark this answers is not in the document]
Comment (PaulRitter) [reply Content.Server/DeviceNetwork/NetworkPacket.cs:18]: i mean you can still loop over the dict & validate it

File: Content.Server/DeviceNetwork/NetworkPacket.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public static NetworkPayload Create(params (string, string)[] data)
             {
                 if(data[index].Item1.Length <= MAX_STRING_SIZE && data[index].Item2.Length <= MAX_STRING_SIZE)
                 {
-                    packet.Add(data[index].Item1, data[index].Item2);
+                    packet.TryAdd(data[index].Item1, data[index].Item2);
```
Comment (PaulRitter) [inline Content.Server/DeviceNetwork/NetworkPacket.cs:25]: i'd at least log an error or smth

**Review by PaulRitter — CHANGES_REQUESTED** [review]

**Comment by PaulRitter** [comment]
xmldocs look good so far btw, forgot to mention

**Comment by juliangiebel** [comment]
I wrote a hacmd page for this:

**Comment by juliangiebel** [comment]
Clicked the wrong button

File: Content.Server/DeviceNetwork/Metadata.cs [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+    public static class Metadata
     {
-        public bool TryParseMetadata<T>(string key, [NotNullWhen(true)] out T data)
+        public static bool TryParseMetadata<T>(this Dictionary<string, object> metadata, string key, [NotNullWhen(true)] out T data)
         {
-            if(TryGetValue(key, out var value) && value is T typedValue)
+            if(metadata.TryGetValue(key, out var value) && value is T typedValue)
             {
                 data = typedValue;
                 return true;
```
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/DeviceNetwork/Metadata.cs:13]: See space-wizards/RobustToolbox#1515

File: Content.Server/GameObjects/Components/ConfigurationComponent.cs [hunk 4]
```diff
@@ -35,6 +36,15 @@ public class ConfigurationComponent : SharedConfigurationComponent, IInteractUsi
 
         private Regex _validation;
 
+        public override void Initialize()
+        {
+            base.Initialize();
+            if (UserInterface != null)
+            {
+                UserInterface.OnReceiveMessage += UserInterfaceOnReceiveMessage;
```
[the remark this answers is not in the document]
Comment (PaulRitter) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs:44]: @juliangiebel why did you resolve this

**Review by PaulRitter — CHANGES_REQUESTED** [review]
i'd like to see this using nullability

File: Content.Server/GameObjects/Components/ConfigurationComponent.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs:44]: Sorry i clicked resolve without clicking on comment first.
Inside the OnAdd method, this got messed up while merging master and now there are both Initialize and OnAdd. I didn't notice that so I'll remove that later today.

[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs:45]: Sorry i clicked resolve without clicking on comment first.
Inside the OnAdd method, this got messed up while merging master and now there are both Initialize and OnAdd. I didn't notice that so I'll remove that later today.

[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs:45]: I removed the OnAdd method, setting the event handler there didn't do anything.

File: Content.Server/GameObjects/Components/ConfigurationComponent.cs [hunk 5]
```diff
@@ -29,12 +30,12 @@ namespace Content.Server.GameObjects.Components
...
     public class ConfigurationComponent : SharedConfigurationComponent, IInteractUsing
     {
-        [ViewVariables] private BoundUserInterface UserInterface => Owner.GetUIOrNull(ConfigurationUiKey.Key);
+        [ViewVariables] private BoundUserInterface? UserInterface => Owner.GetUIOrNull(ConfigurationUiKey.Key);
 
         [ViewVariables]
         private readonly Dictionary<string, string> _config = new();
 
-        private Regex _validation;
+        private Regex _validation = new Regex("^[a-zA-Z0-9 ]*$", RegexOptions.Compiled);
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/ConfigurationComponent.cs:38]: unneeded allocation as it gets overridden by exposedata

**Review by PaulRitter — CHANGES_REQUESTED** [review]

File: Content.Server/GameObjects/Components/ConfigurationComponent.cs (hunk 5 above)
Comment (PaulRitter) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs:38]: mark it as nullable & initialize with null!

Comment (juliangiebel) [reply Content.Server/GameObjects/Components/ConfigurationComponent.cs]: 😅

File: Content.Server/GameObjects/Components/DeviceNetworkConnections/BaseNetworkConnectionComponent.cs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            Connection = _network.Register(DeviceNetID, DeviceNetFrequency, OnReceiveDeviceNetMessage, _receiveAll);
+        }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(ref _receiveAll, "ReceiveAll", false);
+            serializer.DataField(ref _handlePings, "RespondToPings", false);
+            serializer.DataField(ref _pingResponse, "PingResponse", "");
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/DeviceNetworkConnections/BaseNetworkConnectionComponent.cs:47]:
```suggestion
            serializer.DataField(ref _receiveAll, "receiveAll", false);
            serializer.DataField(ref _handlePings, "respondToPings", false);
            serializer.DataField(ref _pingResponse, "pingResponse", "");
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Also nullables as Paul said.

**Comment by PaulRitter** [comment]
closing this due to it going stale, feel free to open it when you got time to work on it again

**Pull request closed without merging**